mono/metadata/metadata.c

```diff
@@ -1083,17 +1083,36 @@ mono_metadata_blob_heap (MonoImage *meta, guint32 index)
 	return meta->heap_blob.data + index;
 }
 
+/**
+ * mono_metadata_blob_heap_null_ok:
+ * \param meta metadata context
+ * \param index index into the blob.
+ * \return an in-memory pointer to the \p index in the Blob heap.
+ * If the Blob heap is empty or missing and index is 0 returns NULL, instead of asserting.
+ */
+const char *
+mono_metadata_blob_heap_null_ok (MonoImage *meta, guint32 index)
```

Comment:
I assume you're adding this because `mono_metadata_blob_heap` is public, but you don't always want the checked version of the function? 

We should probably make the assert in `mono_metadata_blob_heap` clearer/leave a note in case this hits some other part of the code base, since I fear that in the future we'll run into the same issue with some other weird embedder usage.

Reply:
Yea it was partly that, and partly that there are a lot of callers in the runtime and it looked tricky to figure out which ones should be able to deal with a null return value gracefully and which ones definitely expect a non-NULL value.

I added a separate assert for the `index == size == 0` case and a comment with an explanation